compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -832,6 +832,174 @@ struct DistributeLayoutConflictResolutions final
+
+    auto resolutionType =
+        llvm::dyn_cast_or_null<VectorType>(resolutionOp.getResult().getType());
+    if (!resolutionType)
+      return failure();
+    if (!resolutionType.hasStaticShape())
+      return failure();
+    auto paddedShape = SmallVector<int64_t>(resolutionType.getShape());
+    int64_t vectorRank = resolutionType.getRank();
+    paddedShape[vectorRank - 1] *= numSubgroups;
+
+    // Offset and indexing computation such that subgroups can
+    // write and read to shared memory correctly and without conflicts.
+    AffineExpr d0, d1, d2, s0;
+    bindDims(rewriter.getContext(), d0, d1, d2);
+    bindSymbols(rewriter.getContext(), s0);
+    auto indexType = rewriter.getIndexType();
+    Value threadX =
+        rewriter.create<gpu::ThreadIdOp>(loc, indexType, gpu::Dimension::x);
+    Value threadY =
+        rewriter.create<gpu::ThreadIdOp>(loc, indexType, gpu::Dimension::y);
+    Value threadZ =
+        rewriter.create<gpu::ThreadIdOp>(loc, indexType, gpu::Dimension::z);
+    Value flatThreadId = affine::makeComposedAffineApply(
+        rewriter, loc,
+        (d0 + workgroupSize.value()[0] * d1 +
+         (workgroupSize.value()[0] * workgroupSize.value()[1]) * d2),
+        {threadX, threadY, threadZ});
+    Value subgroupOffset = affine::makeComposedAffineApply(
+        rewriter, loc,
+        s0.floorDiv(subgroupSize.value()) *
+            resolutionType.getShape()[vectorRank - 1],
+        {flatThreadId});
+
+    // Create shared memory to store the intermediate from src layout.
+    auto workgroupMemoryAddressSpace = Attribute(gpu::AddressSpaceAttr::get(
+        rewriter.getContext(), gpu::AddressSpace::Workgroup));
+    MemRefType allocType =
+        MemRefType::get(paddedShape, resolutionType.getElementType(),
+                        AffineMap(), workgroupMemoryAddressSpace);
+    auto alloc = rewriter.create<memref::AllocOp>(loc, allocType);
+
+    SmallVector<OpFoldResult> offsets(vectorRank, rewriter.getIndexAttr(0));
+    SmallVector<OpFoldResult> strides(vectorRank, rewriter.getIndexAttr(1));
+    SmallVector<OpFoldResult> shapes = llvm::to_vector(
+        llvm::map_range(resolutionType.getShape(), [&](int64_t dim) {
+          return OpFoldResult(rewriter.getIndexAttr(dim));
+        }));
+    offsets[vectorRank - 1] = subgroupOffset;
+    auto subview = rewriter.create<memref::SubViewOp>(loc, alloc, offsets,
+                                                      shapes, strides);
+
+    // Creating write/trip to shared memory using src layout.
+    Value c0 = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    SmallVector<Value> indices(resolutionType.getRank(), c0);
+    SmallVector<bool> inBounds(vectorRank, true);
+    auto write = rewriter.create<vector::TransferWriteOp>(loc, vector, subview,
+                                                          indices, inBounds);
+    // Insert gpu.barrier
+    rewriter.create<gpu::BarrierOp>(write.getLoc());
```

Comment:
technically, we don't need to add a gpu barrier since the conflict is within a subgroup. fine for now, but good to test maybe.